src/base/types/owl_types_computation_device.ml

```diff
@@ -14,7 +14,7 @@ module type Sig = sig
   type device
   (** TODO *)
 
-  type value
+  type value = ArrVal of A.arr | EltVal of A.elt
```

Comment:
Can we keep this type definition abstract? This is device dependent, if we expose the type definition here, it becomes difficult to implement CG on other devices. 

Based on the two pattern matchings you used, is it possible to use `owl_computation_symbol.is_elt` instead to achieve the same thing?

Reply:
There are also `owl_computation_symbol.is_arr` and `owl_computation_symbol.node_numel` functions, might be helpful?

Reply:
Actually using `owl_computation_symbol.is_elt` in `make_value_block` breaks the `unit_lazy.ml` tests, because it relies on the shape of the node when `make_node` is called ─ in `unit_lazy.ml`, I think the shape is never computed. However would it be reasonable to add an `is_elt : value -> bool` and `is_arr : value -> bool` to `owl_types_computation_device.ml`? The current interface already mentions `elt`'s and `arr`'s and it might be cleaner than using the shape to distinguish them? :)

Reply:
Yes, let's move them to `owl_types_computation_device.ml`. Because `value` is device-dependent, so it should be device's job to tell if it is a scalar or ndarray. Logically, it makes more sense, I agree.